### PR TITLE
RetinaNet object detection (take 2)

### PR DESCRIPTION
**Pull request by fmassa**
This is entirely based on top of the great work from @hgaiser in https://github.com/pytorch/vision/pull/1697

I'm creating a new PR because there are some minor things that could be fixed in that PR but I don't have rights to push to the PR, so in order to move faster I'm creating a new PR but all the history is kept here.

Here is the mAP for the uploaded model:
```
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.364
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.558
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.383
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.193
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.400
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.490
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.315
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.506
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.558
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.386
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.595
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.699
```

### CODE REVIEW

File: torchvision/models/detection/retinanet.py [hunk 1]
```diff
@@ -205,7 +205,7 @@ def compute_loss(self, targets, head_outputs, anchors, matched_idxs):
             target_regression = self.box_coder.encode_single(matched_gt_boxes_per_image, anchors_per_image)
 
             # compute the loss
-            losses.append(det_utils.smooth_l1_loss(
+            losses.append(torch.nn.functional.l1_loss(
```
Comment (hgaiser) [inline torchvision/models/detection/retinanet.py:210]: Why switch it to regular l1 loss?

File: torchvision/models/detection/retinanet.py [hunk 2]
```diff
@@ -100,7 +100,8 @@ def compute_loss(self, targets, head_outputs, matched_idxs):
             foreground_idxs_per_image = matched_idxs_per_image >= 0
             num_foreground = foreground_idxs_per_image.sum()
             # no matched_idxs means there were no annotations in this image
-            if False:#matched_idxs_per_image.numel() == 0:
+            # TODO: enable support for images without annotations that works on distributed
```
Comment (hgaiser) [inline torchvision/models/detection/retinanet.py:103]: What is the problem with images without annotations on distributed?

File: torchvision/models/detection/retinanet.py (hunk 1 above)
Comment (fmassa) [reply torchvision/models/detection/retinanet.py:217]: This gives a 1 mAP improvement on the models, and we were still lagging a bit behind on the mAP compared to detectron2 (which has now adopted the L1 loss by default as well, see https://github.com/facebookresearch/detectron2/commit/b0e268752286b3e643a7af0149cf33f45c5fa64f)

File: torchvision/models/detection/retinanet.py (hunk 2 above)
Comment (fmassa) [reply torchvision/models/detection/retinanet.py:110]: There might be cases where in one GPU no images had annotations, while in the other there were. In this case, there will be part of the computation graph which would not be executed in one GPU, leading to synchronization issues (and even deadlocks).

So for now I'm disabling support for this to move forward, and we will enable it in a later PR (after the release)

Comment (fmassa) [reply torchvision/models/detection/retinanet.py:110]: Note that we are not passing `find_unused_params` to DDP, which you were probably using (because I had to change something else in the `resnet_fpn_backbone` for it to work). With `find_unused_params=True`, this might not be an issue, but I prefer to be on the safer side and let all computation graphs to be the same on every GPU

File: torchvision/models/detection/retinanet.py (hunk 1 above)
Comment (hgaiser) [reply torchvision/models/detection/retinanet.py:217]: Ah nice, wasn't expecting such a difference. What is the mAP that you are getting now? Also, wow, 37.4 mAP. That's impressive.

Comment (fmassa) [reply torchvision/models/detection/retinanet.py:217]: Here is the mAP scores for the model I'll be uploading, with L1 loss:

```
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.364
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.558
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.383
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.193
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.400
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.490
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.315
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.506
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.558
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.386
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.595
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.699
```

**Comment by liminghu** [comment]
> This is entirely based on top of the great work from @hgaiser in #1697
> 
> I'm creating a new PR because there are some minor things that could be fixed in that PR but I don't have rights to push to the PR, so in order to move faster I'm creating a new PR but all the history is kept here.
> 
> Here is the mAP for the uploaded model:
> 
> ```
> IoU metric: bbox
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.364
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.558
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.383
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.193
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.400
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.490
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.315
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.506
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.558
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.386
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.595
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.699
> ```

Thanks a lot. Any ttutorial on how to train the Retinanet on COCO database or other image sets?

**Comment by fmassa** [comment]
@liminghu We have training scripts for all detection models in https://github.com/pytorch/vision/tree/master/references/detection and a finetuning tutorial (for mask rcnn) in https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html
